src/lamp_py/ingestion/convert_gtfs_rt.py

```diff
@@ -405,55 +402,90 @@ def write_local_pq(self, table: pyarrow.Table, local_path: str) -> None:
+
+        :param table: pyarrow Table
+        :param local_path: path to local parquet file
+        """
+        logger = ProcessLogger(
+            "write_local_pq", local_path=local_path, table_rows=table.num_rows
+        )
+        logger.log_start()
+
+        out_ds = self.make_hash_dataset(table, local_path)
+
+        unique_ts_min = pc.min(table.column("feed_timestamp")).as_py() - (
+            60 * 45
+        )
+
+        no_hash_schema = out_ds.schema.remove(
+            out_ds.schema.get_field_index(GTFS_RT_HASH_COL)
         )
 
         with tempfile.TemporaryDirectory() as temp_dir:
-            tmp_pq = os.path.join(temp_dir, "temp.parquet")
-            with pq.ParquetWriter(tmp_pq, schema=out_ds.schema) as writer:
-                for part in partitions:
-                    part_filter = pc.field(part_col) == part
-                    table = pl.from_arrow(out_ds.to_table(filter=part_filter))
-                    writer.write_table(
-                        table.sort(
-                            by=["feed_timestamp"],
+            hash_pq_path = os.path.join(temp_dir, "hash.parquet")
+            upload_path = os.path.join(temp_dir, "upload.parquet")
+            hash_writer = pq.ParquetWriter(hash_pq_path, schema=out_ds.schema)
+            upload_writer = pq.ParquetWriter(upload_path, schema=no_hash_schema)
+
+            partitions = pc.unique(
+                out_ds.to_table(columns=[self.detail.partition_column]).column(
+                    self.detail.partition_column
+                )
+            )
+            for part in partitions:
+                write_table = pyarrow.concat_tables(
+                    [
+                        pl.DataFrame(
+                            out_ds.to_table(
+                                filter=(
+                                    (
+                                        pc.field(self.detail.partition_column)
+                                        == part
+                                    )
+                                    & (
+                                        pc.field("feed_timestamp")
+                                        >= unique_ts_min
+                                    )
+                                )
+                            )
                         )
+                        .sort(by=["feed_timestamp"])
                         .unique(subset=GTFS_RT_HASH_COL, keep="first")
                         .to_arrow()
-                        .cast(out_ds.schema)
-                    )
+                        .cast(out_ds.schema),
+                        out_ds.to_table(
+                            filter=(
+                                (pc.field(self.detail.partition_column) == part)
+                                & (pc.field("feed_timestamp") < unique_ts_min)
+                            )
+                        ),
+                    ]
+                ).sort_by(self.detail.table_sort_order)
```

Comment:
if i'm reading this right, we're doing the deduplication on everything in the out dataset that is less than 45 minutes older than the oldest piece that was ingested in this event loop?

i think it might read a little cleaner if the two tables, newer and older, were explicitly declared and then concatenated. it would take one layer of nesting out.